Add event-flow visualization

williaster opened the pull request:
@mistercrunch @graceguo-supercat 

This PR adds an event flow visualization for analysis of event sequences. The vis is mostly built in [`@data-ui`](https://github.com/williaster/data-ui/tree/chris--event-flow) with [@vx](https://github.com/hshoff/vx) and there is a light wrapper for Superset integration. 

Overall the vis supports the following:
- Takes raw user (or more generically 'entity') event sequences and computes aggregate paths (shown in top panel of vis)
- Allows alignment of event sequences by an arbitrary event **index** and event **type** (e.g., the 2nd click event)
- Allows selection of arbitrary subtrees of the aggregated graph; this filters to the selected subtree in the aggregate panel and shows the corresponding/selected raw event sequences in a bottom panel
- Gives a breakdown of event type counts in the right panel; by clicking on the legend you can filter to / filter out specific types of events.
- X-axis: by default aggregate nodes are positioned according to the mean elapsed time from the previous node, for closely-spaced events you can also position by sequence _number_ (1st, 2nd, 3rd, etc)
- Node sorting: by default nodes are ordered top -> bottom based on high -> low event count meaning most common nodes appear at the top; you can also order by short -> long elapsed time to the next event.

![event-flow-demo](https://user-images.githubusercontent.com/4496521/28036912-9fa47d72-656e-11e7-8ac4-4f34b49cc8d9.gif)

note that builds, etc. will fail until the `@data-ui/event-flow` package is published (which I should get to later today), I just wanted to get some 👀 s on the PR

williaster committed:
[event-flow] add event flow visualizaton type from @data-ui/event-flow.

williaster committed:
[event-flow] update vis thumbnail

williaster commented:
and ere are some pics in Superset:

**Explore controls**
<img height="400" src="https://user-images.githubusercontent.com/4496521/28043576-595a15de-6587-11e7-86b8-25b6c9e063b4.png" />

**Dashboard**
<img height="300" src="https://user-images.githubusercontent.com/4496521/28043485-f24557fa-6586-11e7-9c0a-a741f1948346.png" />

williaster committed:
[event-flow] update row limit label, remove duplicate chart controls

mistercrunch commented:
The image should be square, ideally 1024*1024 to render well

mistercrunch commented:
This is awesome btw! I'll try to review soon but I'm a little swamped at the moment.

mistercrunch commented on superset/assets/visualizations/EventFlow.jsx:
```diff
@@ -0,0 +1,60 @@
+  TS,
+  EVENT_NAME,
+  ENTITY_ID,
+} from '@data-ui/event-flow';
```
what does '@%' do? where does the event-flow code live?

williaster replied:
it's just a name-spaced package on npm, under the ["@data-ui" user](https://www.npmjs.com/~data-ui) but mainly so all the `@data-ui` packages are in the same repo.

itwill live at [https://github.com/williaster/data-ui/packages/event-flow](https://github.com/williaster/data-ui/tree/chris--event-flow) but I haven't quite merged the chart into the master @data-ui repo, adding one more feature. [this is the (massive 🙈) PR](https://github.com/williaster/data-ui/compare/chris--event-flow?expand=1)

williaster committed:
[dependencies] add @data-ui/event-flow 0.0.2

williaster commented:
The package is published now so the build should be fixed (you can demo it [here](https://williaster.github.io/data-ui/?selectedKind=event-flow&selectedStory=hundredUsers&full=0&down=1&left=1&panelRight=0&downPanel=kadirahq%2Fstorybook-addon-actions%2Factions-panel)).

For visual clarity + performance, with larger number of events expected with use in Superset (10,000s) I also added functionality to "trim" nodes in the visualization based on a minimum event count threshold. This means that sequences are aggregated together but "leaf" nodes are hidden from view.

**Before / after trimming**
![event-flow-event-filter](https://user-images.githubusercontent.com/4496521/28106105-59099a76-6697-11e7-8de9-5b6998956ba8.gif)

williaster committed:
[linting] fix multiple imports

williaster committed:
[deps] bump mapbox-gl and react-map-gl to fix build

williaster committed:
[event-flow] bump to 0.0.3 for es2015 + stage-0 babel presets

mistercrunch commented on superset/assets/javascripts/explore/stores/controls.jsx:
```diff
@@ -1273,5 +1273,11 @@ export const controls = {
     description: 'The number of seconds before expiring the cache',
   },
+
+  generic_checkbox: {
```
This is the name that will be use in the saved slice's json. Having something generic here will make it harder if/when trying to generate slice dynamically, and harder to read the json and understand what that maps to. Also references in the backend/frontend code won't be as explicit. Let's use a name that describe the actual feature this triggers.

mistercrunch commented on superset/assets/package.json:
```diff
@@ -56,7 +57,7 @@
     "jsdom": "9.12.0",
     "lodash.throttle": "^4.1.1",
-    "mapbox-gl": "^0.26.0",
+    "mapbox-gl": "^0.27.0",
```
I'm guessing this was an attempt at fixing the `npm install` bug. It's been fixed on master by using `yarn` instead of `npm install` so this is probably not needed anymore. Though if you've tested that the mapbox viz renders alright I'm not against the bump here even though it would belong in another PR ideally.

mistercrunch commented on superset/assets/visualizations/EventFlow.jsx:
```diff
@@ -0,0 +1,60 @@
+  const hasData = json.data && json.data.length > 0;
+
+  // the slice container overflows ~80px in explorer, so we have to correct for this
+  const isExplorer = (/explore/).test(window.location.pathname);
```
not sure if I understand why, but I'm wondering if using `slice.width()` and `slice.height()` would address the 80px issue?

williaster replied:
I tried this to begin with but it doesn't seem to work/couldn't come up with any fix for the css. The panel dom looks like:
```html
<div class="panel">
  <div class="panel-heading">...</div>
  <div class="panel-body">...</div>
</div>
```
`panel-body` is what the slice is in and it seems to have the same height as the parent `panel`. that means it overflows by the height of `panel-heading` :/

mistercrunch reviewed (changes requested):
Sorry for the delay on the review, but this is really amazing and the kind of visualization that is a key "selling" point for Superset. I have a few notes for otherwise LGTM!

I'm curious about your CSS in your module and want to confirm that it's namespace alright so it doesn't spill on other visualization (the same way treemap did on yours).

williaster committed:
[deps] revert mapbox version bumps

williaster committed:
[event-flow] update png, bump to newest version, address reviewer comments, add min event count form.

williaster commented:
@mistercrunch I made the following changes

- removed mapbox version bumps (these broke the build for a while, looks good now 👌)
- updated the thumbnail png to 1024 x 1024
- made the checkbox more specific to event-flow

williaster committed:
[event-flow] pin version

mistercrunch commented:
I fixed the coveralls token in master, please rebase+push

williaster commented:
@mistercrunch done

williaster commented:
let me know if I should add a test for the coverage.

williaster committed:
[event-flow][spec] add test for coveralls

williaster committed:
[event-flow] revert spec

vnnw commented:
Can you add a demo into Superset distribution?

williaster commented:
@vnnw yes this would be a good addition to get people up and going a little more easily.
